### ISSUE TITLE
why dont work timeout?

### ISSUE THREAD

**Issue by alex-qwerty-alex** [phpseclib/phpseclib#1116]
I have a problem:
Сonnection freezes for 60 seceonds, when the function **_get_binary_packet()** called in this place:
 **$raw = stream_get_contents($this->fsock, $this->decrypt_block_size);**
why does't work timeout? it's a bug or а feature?

Maybe you can add **stream_set_timeout**  variable
this will solve my problem

**Comment by terrafrost** [comment]
phpseclib already does [`stream_select`](http://php.net/stream-select) which blocks until the timeout parameter. And it works. Here's a Vagrantfile (1) to demonstrate that the system doesn't have a funky configuration and (2) provide you the means to reproduce my results even if you don't necessarily have the same system:

```
Vagrant.configure("2") do |config|
	config.vm.box = "ubuntu/trusty64"
end
```

PHP code:

```php
$ssh = new SSH2('127.0.0.1', 2200);
$ssh->login('vagrant', 'vagrant');

$start = microtime(true);
$ssh->setTimeout(10);
echo $ssh->read();
$elapsed = microtime(true) - $start;

echo "\ntook $elapsed sec\n";
```
It outputs the banner and the prompt after 10s.

Another example:
```php
$ssh = new SSH2('127.0.0.1', 2200);
$ssh->login('vagrant', 'vagrant');

$start = microtime(true);
$ssh->setTimeout(10);
echo $ssh->exec('ping google.com');
$elapsed = microtime(true) - $start;

echo "\ntook $elapsed sec\n";
```
As before this dumps output in almost exactly 10s.

This was implemented back in 2012:

http://www.frostjedi.com/phpbb3/viewtopic.php?f=46&t=30238
https://github.com/phpseclib/phpseclib/commit/1ab30836a6e145f4c7ae4517de06b2f38a93a1b8

Why `stream_select` instead of `stream_set_timeout`? Keep in mind this was ~5 years ago but...  it looks like `stream_select` would be easier to implement.

As is, it appears to be measuring how much time is passing between the receipt of channel packets. If it takes 1m to receive a channel packet but only 1s for at least one byte of a channel packet to be made available after the last channel packet was received in full I guess the timeout's could be a bit skewed. But I also don't think that's a likely use case.

In http://www.frostjedi.com/phpbb3/viewtopic.php?p=120133#p120133 (from 2010) I write "*I guess the big issue would be that stream_set_timeout() is cumulative*". So by the time you get to 

Why is `stream_select` in `_get_channel_packet` as opposed to `_get_binary_packet`? idk. But I think my earlier point still stands - that long gaps of time between channel packets is the most likely scenario.

Long story short...  I'd be surprised if `setTimeout()`, as currently implemented, didn't solve whatever problem you're having. If it didn't then I'd like to the SSH logs. You can get them by doing `define('NET_SSH2_LOGGING', 2)` and then `echo $ssh->getLog()`. If you want *me* to make changes you have to do a little more to convince me other than just saying "this will solve my problem". If *you* want to make the changes and then submit a PR then I guess the bar isn't gonna be quite as high.

**Comment by alex-qwerty-alex** [comment]
ok, i will try)
log

> <-
> 00000000  53:53:48:2d:32:2e:30:2d:4f:70:65:6e:53:53:48:5f  SSH-2.0-OpenSSH_
> 00000010  36:2e:36:2e:31:70:31:20:44:65:62:69:61:6e:2d:34  6.6.1p1 Debian-4
> 00000020  7e:62:70:6f:37:30:2b:31:0d:0a                    ~bpo70+1..
> 
> ->
> 00000000  53:53:48:2d:32:2e:30:2d:70:68:70:73:65:63:6c:69  SSH-2.0-phpsecli
> 00000010  62:5f:32:2e:30:20:28:6f:70:65:6e:73:73:6c:2c:20  b_2.0 (openssl, 
> 00000020  62:63:6d:61:74:68:29:0d:0a                       bcmath)..
> 

I thinking, this is connection problem.
When connection's socket does not open correctly, timeout option work perfectly.
But if socket open correctly, for **stream_get_contents** this option does not used.
My simple solution, after **fsockopen** аppend

> stream_set_timeout($this->fsock, $this->timeout)

**Comment by terrafrost** [comment]
Sorry for the delay - been busy at work.

Anyway, I assume that's the log when `stream_set_timeout` isn't used? If so then what does it look like when `stream_set_timeout` is used? The next packet should be an incoming NET_SSH2_MSG_KEXINIT. It's not a channel packet so `stream_set_blocking` isn't called. If it works with `stream_set_timeout` then my guess is that you're getting a malformed NET_SSH2_MSG_KEXINIT packet. That the length it reports for the binary packet is a lot larger than the packet itself actually is so `_get_binary_packet` is just waiting and waiting.

The phpseclib logs with `stream_set_timeout` would confirm this. It'd also be interesting to see the "since last" and "network" times.

**Comment by forrest79** [comment]
Hi, I think, that I have issue related to this. I have some long run SSH commands (~minutes). When I set `$timeout = 0`, `exec()` function still return `FALSE` after ~4 minutes. When I set timeout to something big (like 100000) everything is OK. I think, that it's because with `$timeout = 0` is set big timeout to fsockopen, but there is never call `stream_select()` with timeout setting. Call `stream_set_timeout()` after connect could solve this.

**Comment by terrafrost** [comment]
I had hoped to take a look at this this weekend but this weekend didn't really go according to plan. I'll try to take a look at this sometime in the coming week.

Thanks!

**Comment by terrafrost** [comment]
https://github.com/phpseclib/phpseclib/commit/a9c0e2d4300533de468503587ad4021d42393ccd should fix this. I tested it by doing `$ssh->exec('sleep 300; echo test')`.

I'm still not a big fan of `stream_set_timeout` because...  consider this code:

```php
$ssh->setTimeout(30);
echo $ssh->exec('ping 127.0.0.1');
```
With `stream_set_timeout` that'd never return anything - it'd basically be an infinite loop. Unless I set the timeout before each and every `fread` call, which kinda seems inelegant.

I mean, I suppose one could argue that doing `stream_select` before every `fread` call is comparable but idk. `stream_select` just seems more elegant.

**Comment by forrest79** [comment]
Hi, thanks, I can confirm that it works good now.
